transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -906,17 +906,21 @@ private boolean isNotValidPromise(ChannelPromise promise, boolean allowVoidPromi
 
     private AbstractChannelHandlerContext findContextInbound(int mask) {
         AbstractChannelHandlerContext ctx = this;
+        EventExecutor currentExecutor = executor();
         do {
             ctx = ctx.next;
-        } while ((ctx.executionMask & mask) == 0);
+        } while (!ChannelHandlerMask.isInbound(ctx.executionMask)
+            || (ctx.executionMask & mask) == 0 && ctx.executor() == currentExecutor);
         return ctx;
     }
 
     private AbstractChannelHandlerContext findContextOutbound(int mask) {
         AbstractChannelHandlerContext ctx = this;
+        EventExecutor currentExecutor = executor();
         do {
             ctx = ctx.prev;
-        } while ((ctx.executionMask & mask) == 0);
+        } while (!ChannelHandlerMask.isOutbound(ctx.executionMask)
+            || (ctx.executionMask & mask) == 0 && ctx.executor() == currentExecutor);
```

Comment:
this is not correct (and also not the one above) as you will cause a `ClassCastException` later on if the `ChannelHandlerContext` does not wrap an `ChannelOutboundHandler`.